File: ios/RNRectangleScannerView.m
//
//  RNRectangleScannerView.m
//
//  Created by Jake Humphrey on Jan 6, 2020.
//  Copyright (c) 2020 Jake Humphrey. All rights reserved.
//

#import "RNRectangleScannerView.h"

/*!
 Wraps up the camera and rectangle detection code into a simple interface.  Allows you to call start, stop, cleanup, and capture. Also is responsible for deterining how to cache the output images.
 */
@implementation RNRectangleScannerView

- (instancetype)init {
  self = [super init];
  if (self) {
    [self setEnableBorderDetection:YES];
    self._isStopped = TRUE;
    self._cameraIsSetup = FALSE;
    self.capturedQuality = 0.5;
  }

  return self;
}

/*!
 Used to start the camera if it is stopped
 */
- (void) startCamera {
  if ([self _isStopped]) {
    if (![self _cameraIsSetup]) {
      [self setupCameraView];
    }
    [self start];
  }
}

/*!
 Used to stop the camera if it is running
 */
- (void) stopCamera {
  if (![self _isStopped]) [self stop];
}

/*!
 Turns off the torch and stops the camera.
 */
- (void) cleanup {
  [self setEnableTorch: NO];
  [self stopCamera];
  self._cameraIsSetup = NO;
}


/*!
 Called after the camera and session are set up. This lets you check if a camera is found and permission is granted to use it.
 */
- (void)deviceWasSetup:(NSDictionary *)config {
  [super deviceWasSetup:config];
  if (self.onDeviceSetup) {
    self.onDeviceSetup(config);
  }
}


/*!
 Called after the torch state is changed
 */
- (void)torchWasChanged:(BOOL)torchEnabled {
  if (self.onTorchChanged) {
    self.onTorchChanged(@{@"enabled": torchEnabled ? @TRUE : @FALSE});
  }
}

/*!
 Called after the camera and session are set up. This lets you check if a camera is found and permission is granted to use it.
 */
- (void)rectangleWasDetected:(NSDictionary *)detection {
  [super rectangleWasDetected:detection];
  if (self.onRectangleDetected) {
    self.onRectangleDetected(detection);
  }
}

- (void)onErrorOfImageProcessor:(NSDictionary*) errorBody {
  if (self.onErrorProcessingImage) {
    self.onErrorProcessingImage(errorBody);
  }
}

/*!
After capture, the image is stored and sent to the event handler
*/
-(void)onProcessedCapturedImage:(UIImage *)croppedImage initialImage: (UIImage *) initialImage lastRectangleFeature: (CIRectangleFeature *) lastRectangleFeature {
  NSString *dir = [NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES) firstObject];
  NSString *storageFolder = @"RNRectangleScanner";

  dir = [dir stringByAppendingPathComponent:storageFolder];

  NSFileManager *fileManager= [NSFileManager defaultManager];
  NSError *error = nil;
  if(![fileManager createDirectoryAtPath:dir withIntermediateDirectories:YES attributes:nil error:&error]) {
    // An error has occurred, do something to handle it
    NSLog(@"Failed to create directory \"%@\". Error: %@", dir, error);
    [self onErrorOfImageProcessor:@{@"message": @"Failed to create the cache directory"}];
    return;
  }

  NSString *croppedFilePath = [dir stringByAppendingPathComponent:[NSString stringWithFormat:@"C%i.jpeg",(int)[NSDate date].timeIntervalSince1970]];
  NSString *initialFilePath = [dir stringByAppendingPathComponent:[NSString stringWithFormat:@"O%i.jpeg",(int)[NSDate date].timeIntervalSince1970]];
  bool hasCroppedImage = (croppedImage != nil);
  if (!hasCroppedImage) {
    croppedFilePath = initialFilePath;
  }


  if (self.onPictureTaken) {
    self.onPictureTaken(@{
      @"croppedImage": croppedFilePath,
      @"initialImage": initialFilePath
    });
  }

  float quality = 0.5;
  if (self.capturedQuality) {
    quality = self.capturedQuality;
  }
  @autoreleasepool {
    if (hasCroppedImage) {
      NSData *croppedImageData = UIImageJPEGRepresentation(croppedImage, quality);
      if (![croppedImageData writeToFile:croppedFilePath atomically:YES]) {
        NSMutableDictionary *errorBody = [[NSMutableDictionary alloc] init];
        [errorBody setValue:@"Failed to write cropped image to cache" forKey:@"message"];
        [errorBody setValue:croppedFilePath forKey:@"filePath"];
        [self onErrorOfImageProcessor:errorBody];
        return;
      }
    }

    NSData *initialImageData = UIImageJPEGRepresentation(initialImage, quality);
    if (![initialImageData writeToFile:initialFilePath atomically:YES]) {
      NSMutableDictionary *errorBody = [[NSMutableDictionary alloc] init];
      [errorBody setValue:@"Failed to write original image to cache" forKey:@"message"];
      [errorBody setValue:initialFilePath forKey:@"filePath"];
      [self onErrorOfImageProcessor:errorBody];
      return;
    }

    if (self.onPictureProcessed) {
      self.onPictureProcessed(@{
        @"croppedImage": croppedFilePath,
        @"initialImage": initialFilePath
      });
    }
  }
}

/*!
 Captures the current frame and sends the processed image(s) to the on picture taken callback.
 */
- (void) capture {
  [self captureImageLater];
}


@end

File: android/src/main/java/com/rectanglescanner/views/RNRectangleScannerView.java
package com.rectanglescanner.views;

import android.app.Activity;
import android.content.Context;
import android.util.Log;
import android.widget.FrameLayout;

import com.rectanglescanner.R;
import com.rectanglescanner.helpers.CapturedImage;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeMap;

import org.opencv.core.Core;
import org.opencv.core.CvType;
import org.opencv.core.Mat;
import org.opencv.core.MatOfInt;
import org.opencv.imgcodecs.Imgcodecs;

import java.util.UUID;
import java.io.File;
import java.util.ArrayList;

/**
  Created by Jake on Jan 6, 2020.

  Wraps up the camera and rectangle detection code into a simple interface.
  Allows you to call start, stop, cleanup, and capture. Also is responsible
  for deterining how to cache the output images.
*/
public class RNRectangleScannerView extends RectangleDetectionController {
    private String cacheFolderName = "RNRectangleScanner";
    private double capturedQuality = 0.5;

    //================================================================================
    // Setup
    //================================================================================

    public RNRectangleScannerView(Context context, Integer numCam, Activity activity, FrameLayout frameLayout) {
        super(context, numCam, activity, frameLayout);
    }

    private MainView parentView = null;

    public void setParent(MainView view) {
      this.parentView = view;
    }

    /**
    Sets the jpeg quality of the output image
    */
    public void setCapturedQuality(double quality) {
      this.capturedQuality = quality;
    }

    /**
    Call to capture an image
    */
    public void capture() {
      captureImageLater();
    }

    /**
     Called after a picture was captured
     */
    private void pictureWasTaken(WritableMap pictureDetails) {
      Log.d(TAG, "picture taken");
      this.parentView.pictureWasTaken(pictureDetails);
    }

    /**
     Called after a picture was captured and finished processing
     */
    private void pictureWasProcessed(WritableMap pictureDetails) {
      Log.d(TAG, "picture processed");
      this.parentView.pictureWasProcessed(pictureDetails);
    }

    /**
    Called if the picture faiiled to be captured
    */
    private void pictureDidFailToProcess(WritableMap errorDetails) {
      Log.d(TAG, "picture failed to process");
      this.parentView.pictureDidFailToProcess(errorDetails);
    }

    /**
     Called after the torch/flash state was changed
     */
    @Override
    protected void torchWasChanged(boolean torchEnabled) {
      Log.d(TAG, "torch changed");
      this.parentView.torchWasChanged(torchEnabled);
    }

    /**
     Called after the camera and session are set up. This lets you check if a
     camera is found and permission is granted to use it.
     */
    @Override
    protected void deviceWasSetup(WritableMap config) {
      Log.d(TAG, "device setup");
      this.parentView.deviceWasSetup(config);
    }


    /**
     Called after a frame is processed and a rectangle was found
     */
    @Override
    public void rectangleWasDetected(WritableMap detection) {
      this.parentView.rectangleWasDetected(detection);
    }


    /**
     After an image is captured and cropped, this method is called
     */
    @Override
    public void onProcessedCapturedImage(CapturedImage capturedImage) {
      WritableMap pictureWasTakenConfig = new WritableNativeMap();
      WritableMap pictureWasProcessedConfig = new WritableNativeMap();
      String croppedImageFileName = null;
      String originalImageFileName = null;
      boolean hasCroppedImage = (capturedImage.processed != null);
      try {
        originalImageFileName = generateStoredFileName("O");
        if (hasCroppedImage) {
          croppedImageFileName = generateStoredFileName("C");
        } else {
          croppedImageFileName = originalImageFileName;
        }
      } catch(Exception e) {
        WritableMap folderError = new WritableNativeMap();
        folderError.putString("message", "Failed to create the cache directory");
        pictureDidFailToProcess(folderError);
        return;
      }

      pictureWasTakenConfig.putString("croppedImage", "file://" + croppedImageFileName);
      pictureWasTakenConfig.putString("initialImage", "file://" + originalImageFileName);
      pictureWasProcessedConfig.putString("croppedImage", "file://" + croppedImageFileName);
      pictureWasProcessedConfig.putString("initialImage", "file://" + originalImageFileName);
      pictureWasTaken(pictureWasTakenConfig);

      if (hasCroppedImage && !this.saveToDirectory(capturedImage.processed, croppedImageFileName)) {
        WritableMap fileError = new WritableNativeMap();
        fileError.putString("message", "Failed to write cropped image to cache");
        fileError.putString("filePath", croppedImageFileName);
        pictureDidFailToProcess(fileError);
        return;
      }
      if (!this.saveToDirectory(capturedImage.original, originalImageFileName)) {
        WritableMap fileError = new WritableNativeMap();
        fileError.putString("message", "Failed to write original image to cache");
        fileError.putString("filePath", originalImageFileName);
        pictureDidFailToProcess(fileError);
        return;
      }

      pictureWasProcessed(pictureWasProcessedConfig);
      capturedImage.release();
      Log.d(TAG, "Captured Images");
    }

    private String generateStoredFileName(String name) throws Exception {
      String folderDir = this.mContext.getCacheDir().toString();
      File folder = new File( folderDir + "/" + this.cacheFolderName);
      if (!folder.exists()) {
          boolean result = folder.mkdirs();
          if (result) {
            Log.d(TAG, "wrote: created folder " + folder.getPath());
          } else {
            Log.d(TAG, "Not possible to create folder");
            throw new Exception("Failed to create the cache directory");
          }
      }
      return folderDir + "/" + this.cacheFolderName + "/" + name + UUID.randomUUID() + ".png";
    }

    /**
    Saves a file to a folder
    */
    private boolean saveToDirectory(Mat doc, String fileName) {
        Mat endDoc = new Mat(doc.size(), CvType.CV_8UC4);
        doc.copyTo(endDoc);
        Core.flip(doc.t(), endDoc, 1);
        ArrayList<Integer> parameters = new ArrayList();
        parameters.add(Imgcodecs.CV_IMWRITE_JPEG_QUALITY);
        parameters.add((int)(this.capturedQuality * 100));
        MatOfInt par = new MatOfInt();
        par.fromList(parameters);
        boolean success = Imgcodecs.imwrite(fileName, endDoc, par);

        endDoc.release();

        return success;
    }

}
